homeassistant/components/drop_connect/coordinator.py

```diff
@@ -0,0 +1,184 @@
+"""DROP device data update coordinator object."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.json import JSON_DECODE_EXCEPTIONS, json_loads
+
+from .const import (
+    CONF_DATA_TOPIC,
+    CONF_DEVICE_DESC,
+    CONF_DEVICE_ID,
+    CONF_DEVICE_NAME,
+    CONF_DEVICE_OWNER_ID,
+    CONF_DEVICE_TYPE,
+    CONF_HUB_ID,
+    DEV_HUB,
+    DOMAIN,
+    KEY_STATUS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DROPDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """DROP device object."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, unique_id: str) -> None:
+        """Initialize the device."""
+        super().__init__(hass, _LOGGER, name=f"{DOMAIN}-{unique_id}")
+        self._model: str = f"{self.config_entry.data[CONF_DEVICE_DESC]} on hub {self.config_entry.data[CONF_HUB_ID]}"
+        if self.config_entry.data[CONF_DEVICE_TYPE] == DEV_HUB:
+            self._model = f"Hub {self.config_entry.data[CONF_HUB_ID]}"
+        self._manufacturer: str = "Chandler Systems, Inc."
+        self._device_name: str = self.config_entry.data[CONF_DEVICE_NAME]
+        self._device_information: dict[str, Any] = {}
+
+    async def drop_message_received(
+        self, topic: str, payload: str, qos: int, retain: bool
+    ) -> None:
+        """Process a received MQTT message."""
+        topic_root = self.config_entry.data[CONF_DATA_TOPIC].removesuffix("/#")
```

Comment:
All MQTT parsing should happen in a 3rd party library. Look at the tasmota integration for an example.

Reply:
@MartinHjelmare  - Would it be acceptable to defer a restructure to move parsing and callbacks to a new 3rd party library to an improvement PR in Q1 2024?  A release of firmware and apps for the DROP product line is scheduled for early January, and we'd like to have an integration presence in HA with the January HA release.  Internal schedule pressures make it unlikely that I will have time to do that restructure before a holiday break at the end of December.

I am not opposed to making the change, I'm just hoping I can do it next month while still having the `drop_connect` integration part of the January HA release.

Reply:
Sorry, we can't do that. Using a 3rd party library is a foundational requirement of integrations integrating devices and services.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Reply:
Thanks for the response.  I'll try to wedge it into the schedule.